xbmc/cores/AudioEngine/Sinks/AESinkPULSE.cpp

```diff
@@ -423,6 +449,16 @@ bool CAESinkPULSE::Initialize(AEAudioFormat &format, std::string &device)
 
   pa_threaded_mainloop_lock(m_MainLoop);
 
+  {
+    // Register Callback for Sink changes
+    CSingleLock lock(m_sec);
+    pa_context_set_subscribe_callback(m_Context, SinkChangedCallback, this);
```

Comment:
@fritsch this should run independently from an active sink like it is done for ALSA. if there is no active pa sink , AE is not notified about new or removed sinks.

Reply:
If AE becomes active and tries to play something it won't open and then fallback to "default" device right? Therefore I think that's why no one complained yet.

Most users use "default" in PA, therefore it will always open and play again.

Reply:
> fallback to "default" device right?

Yes, I think so. I wanted to use PA as additional driver for airplay sinks only.  Problem is that those sinks won't show up in the list of audio devices.

Reply:
They don't show up in general or only if a hotplug happens. 

Could you give me: pactl list sinks

please when they are connected?

Reply:
They show up in pavucontrol but not in AE for the reason given: if there is no active PA sink, there is no callback for sinks added/removed.
If PA is used as the only driver and change happens while there is an active PA sink, DeviceChange is triggered and the sinks are added.

Solution: register the callback in the context of the driver not in the context of the sink.

Reply:
To be honest, I really don't like what ALSA is doing:
```
#if HAVE_LIBUDEV
CALSADeviceMonitor CAESinkALSA::m_deviceMonitor; // ARGH
#endif

#if !defined(HAVE_X11)
CALSAHControlMonitor CAESinkALSA::m_controlMonitor; // ARGH
#endif
```
is a design flaw by excellence-> see the "ARGH" of the original author. AE has no real "driver context" in my opinion - the only specific context it has, is a Sink implementation, which is encapsulated away via ActiveAESink's interface to its m_sink <- here we can speak "pulseish".

What about adding an interface so that something more "upper" e.g. AE which is used for DeviceChanged anyways or ActiveAESink maintaining this driver?

One could create that and transfer ownership to that? It can then thoroughly shutdown it when shutting down, etc.?

Reply:
I agree, the ALSA solution is not ideal. It was done before divers did register. I would create the driver context in CAESinkPULSE::Register. Create a cleanup function, register it, and delete the context there.